Firestore/core/src/firebase/firestore/api/document_reference.mm

```diff
@@ -52,8 +55,21 @@
 using model::Precondition;
 using model::ResourcePath;
 using util::MakeNSError;
+using util::Status;
 using util::StatusOr;
 
+DocumentReference::DocumentReference(Firestore* firestore,
+                                     model::ResourcePath path)
+    : firestore_{firestore} {
+  if (path.size() % 2 != 0) {
```

Comment:
I'm not sure about where this error check should be. Right now, this duplicates the check in [`FIRDocumentReference`](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/API/FIRDocumentReference.mm#L210-L214). It's used in `FIRFirestore`:
```cpp
- (FIRDocumentReference *)documentWithPath:(NSString *)documentPath {
  if (!documentPath) {
    FSTThrowInvalidArgument(@"Document path cannot be nil.");
  }
  if ([documentPath containsString:@"//"]) {
    FSTThrowInvalidArgument(@"Invalid path (%@). Paths must not contain // in them.", documentPath);
  }

  DocumentReference documentReference = _firestore->GetDocument(util::MakeString(documentPath));
  return [FIRDocumentReference referenceWithReference:std::move(documentReference) firestore:self];
}
```
Here, I'm trying to delegate as much as possible to core `Firestore`, so it doesn't make sense to use the static constructor of `FIRDocumentReference` that does the validation. On the other hand, I think we discussed that error checking should mostly be in the wrappers, not core. Also, it seems a little weird that the error checking of the argument here is split among Firestore and document reference... Perhaps move the check to `FIRFirestore documentWithPath`?

Reply:
This seeming duplication is necessary, unfortunately.

One is validating that the user has supplied valid arguments to a public function. The second is asserting that the value is sane. We can't use HARD_ASSERT for argument validation and we can throw invalid argument on an internal failure.